### DISCUSSION TITLE
Equality of nested readonly record structs?

### DISCUSSION THREAD

**Discussion by Simonl9l** [dotnet/csharplang#5777] [Language Ideas] [answered]
If you have say:

```C#
public readonly record struct RecA(string Name, RecB[] RecBs);

public readonly record struct RecB(string Name);
```
and set up
```C#
var a1 = new RecA { Name = "abc", RecBs = new[] { new RecB{ Name = "abc1" }, new RecB{ Name = "abc2" } } };
var a2 = new RecA { Name = "abc", RecBs = new[] { new RecB{ Name = "abc1" }, new RecB{ Name = "abc2" } } };
```
then should not both `res1` & `res2`(below)  == `true`?
```C#
var res1 = a1.Equal(a2);
var res2 = a1 == a2; 
``` 
As a related aside a similar set of struct's is being deserialized from Json using the `Text.Json` parser, hence the desire to use nested records, and then be able to compare the item in said array.

**Comment by Joe4evr** [comment, marked as answer]
This is because one of the fields is an array type and the two instances each have a different array. [The feature spec states](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/record-structs.md#equality-members):
> The synthesized `Equals(R)` returns `true` if and only if for each instance field `fieldN` in the record struct
the value of `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type is `true`.

Since arrays don't do deep equality by default (only reference equality), the two records are considered unequal. [This can be shown by hoisting the array to a local variable and observing that sharing the same array instance makes the two instances equal.](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChisBmAAmQFMIATAexgBsBPK6gYweTooGcEaTggoAlLgEEAFFjwAGCgDkIYanHFcAQgG0Aups5b+ASgDcpSjXpM2HbrwFDUIw1tkLlq6ueKk8AGwUuMF4AOzEAN7EFLGhQVgALBQAstAw0iYxcdFEcfkUAG4QyBTA/BQAvBQw1ADu+hSRNfVuzSpqVRQARBDAnHjdFAC+GrV1bV6d1b39OEPDIxZ5BbHFpRB4XeOGkk1T1F2znN0aEkYV1eVL2avrFBA4263ne+3eR30nZ9qXZRXDZa3ApyACc0k2ADoAKIAR1QEGY/AhOBMvhWqzBEK2lWqj3R+WGxCJRCAA==)

Another option is to manually specify your equality method, like this:
```cs
public readonly record struct RecA(string Name, RecB[] RecBs)
{
    public bool Equals(RecA other)
    {
        return EqualityComparer<string>.Default.Equals(this.Name, other.Name)
            && this.RecBs.SequenceEqual(other.RecBs);
    }
};
```
[Related discussion from literally yesterday.](https://github.com/dotnet/csharplang/discussions/5767)

**Reply by Simonl9l** [reply]
Thanks @Joe4evr for the quick reply - what you highlight makes sense! There is limited docs (did not pick up your excerpt) and examples of nested  record structs, and I'd has the assumption that the default implementation of the Equality would (given we're taking value types) do a deep equality - and this will possibly be a source of ongoing confusion with out more example in the docs to call out this subtlety.

**Reply by Joe4evr** [reply]
> I'd has the assumption that the default implementation of the Equality would (given we're taking value types) do a deep equality

No, [regular structs also don't do deep equality (of a reference type field) by default](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChisBmAAmQFMIATAexgBsBPCgZwTQGMEKAyoSIBvYhQkVyUvAAYKAOQhhqFERQDm1BAG4KsKLooBfcZOkCcAbQC6gnBzWbteg0dNEP0mvSZtO3Kh89sRiRJJSlFhyisqq6lpGbnoeHqR4AGxSODIA7KFmEtFZWAAsFACy0DAAFACUhWqNEQBuEMgUEHgUALwUMNQA7oLd6koqvRQARBDAPFNw9o59A4O2Tqv2TuOqfTNzeFMmi5uW23GT+zw4R8YmJjrNkm0dEDkrQyPnE3uz84uWZb9IbrdSnHJjC6/A63E6fM6Qn7TP43e53YyPEjhCLmPAAThqXQAdABRACOqAgzA4hJwdTqmIiqWIQA), so this is not new.

**Reply by RikkiGibson** [reply]
Should the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#value-equality) be revised to make this more clear?

**Comment by Simonl9l** [comment]
@Joe4evr Acknowledged and understood but we’re talking about `readonly record struct` here and a big (perceived) value add would be that they’re immutable value types - that act that way in all regards. It’s understood that nesting by arrays or collections is the issue. If it were simple 1:1 nesting it would work!

At this point I’d suggest this discussion is asked and answered and we can leave this here - and do appreciate the help!

@RikkiGibson in terms if documentation, and the link you shared on equality, it seems as I read it all to not call out collections and the need as @Joe4evr pointed out to provide an  `Equals` Impl for the record that calls the requisite `SequenceEquals` to include the collection members. 

Whilst I get we should know that it avoids assumptions for those that don’t yet!

Thanks again all!
